dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/ListSink.kt

```diff
@@ -0,0 +1,202 @@
+package org.jetbrains.kotlinx.dataframe.impl.io
+
+import io.deephaven.csv.parsers.DataType
+import io.deephaven.csv.parsers.DataType.BOOLEAN_AS_BYTE
+import io.deephaven.csv.parsers.DataType.BYTE
+import io.deephaven.csv.parsers.DataType.CHAR
+import io.deephaven.csv.parsers.DataType.DATETIME_AS_LONG
+import io.deephaven.csv.parsers.DataType.DOUBLE
+import io.deephaven.csv.parsers.DataType.FLOAT
+import io.deephaven.csv.parsers.DataType.INT
+import io.deephaven.csv.parsers.DataType.LONG
+import io.deephaven.csv.parsers.DataType.SHORT
+import io.deephaven.csv.parsers.DataType.STRING
+import io.deephaven.csv.parsers.DataType.TIMESTAMP_AS_LONG
+import io.deephaven.csv.sinks.Sink
+import io.deephaven.csv.sinks.SinkFactory
+import io.deephaven.csv.sinks.Source
+import kotlinx.datetime.toKotlinLocalDateTime
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import kotlin.time.Duration.Companion.nanoseconds
+
+internal interface SinkSource<T : Any> :
+    Sink<T>,
+    Source<T>
+
+/**
+ * Implementation of Deephaven's [Sink] and [Source] that stores data in an [ArrayList].
+ *
+ * The implementation is based on [Writing Your Own Data Sinks](https://github.com/deephaven/deephaven-csv/blob/main/ADVANCED.md).
+ *
+ * If we ever store column data unboxed / primitively, this needs to be modified.
```

Comment:
Actually, we could also create a `ListSink` for each primitive type and `.asList()` the result. DataFrame can handle such columns. However, this could slow down processing after reading (as all values visited will need to be boxed) and potentially postpone the OOM exception. Unless the user filters first maybe...